### PR TITLE
[ML] Fix ML memory tracker lockup when inner step fails

### PR DESCRIPTION
**Pull request by droberts195**
When the ML memory tracker is refreshed and a refresh is
already in progress the idea is that the second and
subsequent refresh requests receive the same response as
the currently in progress refresh.

There was a bug that if a refresh failed then the ML
memory tracker's view of whether a refresh was in progress
was not reset, leading to every subsequent request being
registered to receive a response that would never come.

This change makes the ML memory tracker pass on failures
as well as successes to all interested parties and reset
the list of interested parties so that further refresh
attempts are possible after either a success or failure.

This fixes problem 1 of #44156

### CODE REVIEW

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
run elasticsearch-ci/1

^ docs test failure

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java [hunk 1]
```diff
@@ -299,7 +299,18 @@ void refresh(PersistentTasksCustomMetaData persistentTasks, ActionListener<Void>
                 }
                 fullRefreshCompletionListeners.clear();
             }
-        }, onCompletion::onFailure);
+        },
+        e -> {
+            synchronized (fullRefreshCompletionListeners) {
+                assert fullRefreshCompletionListeners.isEmpty() == false;
+                for (ActionListener<Void> listener : fullRefreshCompletionListeners) {
+                    listener.onFailure(e);
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java]: I was going to comment on how we don't signal `onCompletion` any more, but then I saw line 286. This is definitely a sneaky bug.

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java (hunk 1 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java]: Yes, as well as making it impossible to retry the bug also meant that only 1 of the queuing listeners got notified of failures.  The others didn't receive any notification at all.

**Pull request merged**